ehrapy/logger.py

```diff
@@ -0,0 +1,199 @@
+from typing import Optional, IO
+
+HINT = (INFO + DEBUG) // 2
+logging.addLevelName(HINT, "HINT")
+
+
+class _RootLogger(logging.RootLogger):
+    def __init__(self, level):
+        super().__init__(level)
+        self.propagate = False
+        _RootLogger.manager = logging.Manager(self)
+
+    def log(
+        self,
+        level: int,
+        msg: str,
+        *,
+        extra: Optional[dict] = None,
+        time: datetime = None,
+        deep: Optional[str] = None,
+    ) -> datetime:
+        from ehrapy import settings
+
+        now = datetime.now(timezone.utc)
+        time_passed: timedelta = None if time is None else now - time
+        extra = {
+            **(extra or {}),
+            "deep": deep if settings.verbosity.level < level else None,
+            "time_passed": time_passed,
+        }
+        super().log(level, msg, extra=extra)
+        return now
+
+    def critical(self, msg, *, time=None, deep=None, extra=None) -> datetime:
+        return self.log(CRITICAL, msg, time=time, deep=deep, extra=extra)
+
+    def error(self, msg, *, time=None, deep=None, extra=None) -> datetime:
+        return self.log(ERROR, msg, time=time, deep=deep, extra=extra)
+
+    def warning(self, msg, *, time=None, deep=None, extra=None) -> datetime:
+        return self.log(WARNING, msg, time=time, deep=deep, extra=extra)
+
+    def info(self, msg, *, time=None, deep=None, extra=None) -> datetime:
+        return self.log(INFO, msg, time=time, deep=deep, extra=extra)
+
+    def hint(self, msg, *, time=None, deep=None, extra=None) -> datetime:
+        return self.log(HINT, msg, time=time, deep=deep, extra=extra)
+
+    def debug(self, msg, *, time=None, deep=None, extra=None) -> datetime:
+        return self.log(DEBUG, msg, time=time, deep=deep, extra=extra)
+
+
+class _LogFormatter(logging.Formatter):
+    def __init__(self, fmt="{levelname}: {message}", datefmt="%Y-%m-%d %H:%M", style="{"):
+        super().__init__(fmt, datefmt, style)
+
+    def format(self, record: logging.LogRecord):
+        format_orig = self._style._fmt
+        if record.levelno == INFO:
+            self._style._fmt = "{message}"
+        elif record.levelno == HINT:
+            self._style._fmt = "--> {message}"
+        elif record.levelno == DEBUG:
+            self._style._fmt = "    {message}"
+        if record.time_passed:
+            # strip microseconds
+            if record.time_passed.microseconds:
+                record.time_passed = timedelta(seconds=int(record.time_passed.total_seconds()))
+            if "{time_passed}" in record.msg:
+                record.msg = record.msg.replace("{time_passed}", str(record.time_passed))
+            else:
+                self._style._fmt += " ({time_passed})"
+        if record.deep:
+            record.msg = f"{record.msg}: {record.deep}"
+        result = logging.Formatter.format(self, record)
+        self._style._fmt = format_orig
+        return result
+
+
+_DEPENDENCIES_NUMERICS = [
```

Comment:
We should either get rid of this here or in the meta_information.py script -> https://github.com/theislab/ehrapy/blob/development/ehrapy/core/meta_information.py

I think that for you now it is easier if you delete the duplications in the logger.py script and add any missing things (e.g. `_DEPENDENCIES_PLOTTING` etc to the existing meta_information.py script.